src/ucs/memory/rcache.inl

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (C) 2022, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCS_REG_CACHE_INL_
+#define UCS_REG_CACHE_INL_
+
+#include "rcache_int.h"
+
+#include <ucs/arch/atomic.h>
+
+static inline int ucs_rcache_region_test(ucs_rcache_region_t *region, int prot)
+{
+    return (region->flags & UCS_RCACHE_REGION_FLAG_REGISTERED) &&
+           ucs_test_all_flags(region->prot, prot);
+}
+
+static inline ucs_status_t
+ucs_rcache_get_locked(ucs_rcache_t *rcache, void *address, size_t length,
+                      int prot, void *arg, ucs_rcache_region_t **region_p)
+{
+    ucs_pgt_addr_t start = (uintptr_t)address;
+    ucs_pgt_region_t *pgt_region;
+    ucs_rcache_region_t *region;
+
+    if (ucs_queue_is_empty(&rcache->inv_q)) {
+        pgt_region = UCS_PROFILE_CALL(ucs_pgtable_lookup, &rcache->pgtable,
+                                      start);
+        if (ucs_likely(pgt_region != NULL)) {
+            region = ucs_derived_of(pgt_region, ucs_rcache_region_t);
+            if (((start + length) <= region->super.end) &&
+                ucs_rcache_region_test(region, prot))
+            {
+                ucs_rcache_region_hold(rcache, region);
+                if (region->lru_flags & UCS_RCACHE_LRU_FLAG_IN_LRU) {
+                    ucs_list_del(&region->lru_list);
+                    region->lru_flags &= ~UCS_RCACHE_LRU_FLAG_IN_LRU;
+                }
+                *region_p = region;
+                return UCS_OK;
+            }
+        }
+    }
+
+    return UCS_PROFILE_CALL(ucs_rcache_create_region, rcache, address, length,
+                            prot, arg, region_p);
+}
+
+static inline void
+ucs_rcache_put_locked(ucs_rcache_t *rcache, ucs_rcache_region_t *region)
+{
+    if (ucs_likely(ucs_atomic_fsub32(&region->refcount, 1) != 1)) {
```

Comment:
1. maybe reuse ucs_rcache_region_lru_put_internal
2. missing ucs_rcache_region_lru_put
3. can replace atomic operations by regular math (bus locking is expensive)